8.0 PXC-3191 : cluster cannot update wsrep_* tables if in super_read_only

kennt-percona opened the pull request:
Issue
If super-read-only=ON, the system cannot start up correctly. There are
two problems (1) we cannot modify the wsrep_xxx system tables and (2) the
SST donor cannot create/modify the mysql.pxc.sst.user to perform the
backup

Solution
(1) skip the readonly checks if we are modifying a wsrep_xxx table
(2) skip the readonly checks if we are running as the PXC internal
    session user.

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_super_read_only.test:
```diff
@@ -0,0 +1,24 @@
+--echo [connection node_2]
+--let $restart_parameters = "restart: --super-read-only"
+--source include/restart_mysqld.inc
+--source include/wait_until_connected_again.inc
```
This is redundant as `include/restart_mysqld.inc` does poll the server until it becomes online.

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_super_read_only.test:
```diff
@@ -0,0 +1,24 @@
+# Test handling of super-read-only on server startup
+# This test will shutdown node2 and then restart it with
```
I guess is better we add a scenario where the donor has `super-read-only` set and the joiner tries to join the cluster.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -344,6 +344,12 @@ PSI_file_info all_galera_files[] = {
 typedef std::vector<void *> wsrep_psi_key_vec_t;
 static wsrep_psi_key_vec_t wsrep_psi_key_vec;
 
+LEX_CSTRING PXC_INTERNAL_SESSION_USER =
```
The output of clang-format-8 on all the updated files seems to be differing from the current patch. Can you check this?
```
-LEX_CSTRING PXC_INTERNAL_SESSION_USER =
-    {STRING_WITH_LEN("mysql.pxc.internal.session")};
-LEX_CSTRING PXC_INTERNAL_SESSION_HOST =
-    {STRING_WITH_LEN("localhost")};
-
+LEX_CSTRING PXC_INTERNAL_SESSION_USER = {
+    STRING_WITH_LEN("mysql.pxc.internal.session")};
+LEX_CSTRING PXC_INTERNAL_SESSION_HOST = {STRING_WITH_LEN("localhost")};

```

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_sst_xtrabackup-v2_super_read_only.test:
```diff
@@ -1,26 +1,97 @@
+# --------------------------------
+
+# restart the server with super-read-only=ON (no SST)
+--echo "# restarting node2 (no SST)"
+--let $restart_hide_args=1
+--let $restart_parameters = "restart: --defaults-file=$NODE2_CONFIG_FILE"
+--source include/restart_mysqld.inc
+--source include/wait_until_connected_again.inc
```
Wouldn't just this be enough? Is it really required to create a temporary config file?
```
+ --echo "# restarting node2 (no SST)"
+ --let $restart_parameters = "restart: --super-read-only=ON"
+ --source include/restart_mysqld.inc
```

kennt-percona replied:
see the comment below (it matters more for the SST process).

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_sst_xtrabackup-v2_super_read_only.test:
```diff
@@ -1,26 +1,97 @@
---exec sed -i '/^\[mysqld.2\]/a super_read_only=ON' $MYSQLTEST_VARDIR/my.cnf
+
+# restart the server with super-read-only=ON (SST)
+--echo "# restarting node2 (with SST)"
+--let $restart_hide_args=1
+--let $start_mysqld_params = "--defaults-file=$NODE2_CONFIG_FILE"
 --let $_expect_file_name= $MYSQLTEST_VARDIR/tmp/mysqld.2.expect
 --source include/start_mysqld.inc
+--source include/wait_until_connected_again.inc
```
Same as above.
```
+ # restart the server with super-read-only=ON (SST)
+ --echo "# restarting node2 (with SST)"
+ --let $restart_parameters = "restart: --super-read-only=ON"
  --let $_expect_file_name= $MYSQLTEST_VARDIR/tmp/mysqld.2.expect
  --source include/start_mysqld.inc
```

kennt-percona replied:
The SST script does not have access to all of the command-line parameters, so it reads from the config file.  Thus we need the config file to test how the SST handles the super-read-only option.

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_sst_xtrabackup-v2_super_read_only.test:
```diff
@@ -1,26 +1,97 @@
+
+--let $restart_parameters = "restart:"
+--source include/restart_mysqld.inc
+--source include/wait_until_connected_again.inc
```
This is redundant. We can avoid including `wait_until_connected_again.inc` after using either the `start_mysqld.inc` and `restart_mysqld.inc`, as the same check is already done inisde the inc files.

kamil-holubicki commented on sql/table.h:
```diff
@@ -3959,6 +3965,20 @@ inline bool is_user_table(TABLE *table) {
+          !my_strcasecmp(system_charset_info, WSREP_DB_NAME.str,
+                         db_name) &&
+          WSREP_TABLE_PREFIX.length < table_len &&
+          !native_strncasecmp(table_name, WSREP_TABLE_PREFIX.str,
```
why we use strcmp and strNcmp? 
I would prefer using strNcmp, but I see is_perfschema_db uses strcmp (so probably it is ensured by the caller that argument is the proper, null-terminated string).

At least I would stick to one of them if there is no reason for using different comparison functions.

kamil-holubicki commented on sql/auth/sql_authorization.cc:
```diff
@@ -1940,12 +1940,20 @@ bool check_readonly(THD *thd, bool err_if_readonly) {
+
+  /* Ignore readonly for PXC internal user */
+  if (sctx->user().str != nullptr &&
+      !strcmp(sctx->user().str, PXC_INTERNAL_SESSION_USER.str) &&
```
Suggestion: Is it safe to use strcmp here? Maybe strNcmp would be better?

kennt-percona committed:
PXC-3191 : cluster cannot update wsrep_* tables if in super_read_only

Issue
If super-read-only=ON, the system cannot start up correctly. There are
two problems (1) we cannot modify the wsrep_xxx system tables and (2) the
SST donor cannot create/modify the mysql.pxc.sst.user to perform the
backup

Solution
(1) skip the readonly checks if we are modifying a wsrep_xxx table
(2) skip the readonly checks if we are running as the PXC internal
    session user.

kamil-holubicki reviewed (approved):
lgtm

kennt-percona commented:
https://pxc.cd.percona.com/job/pxc80-test-mtr/230/

venkatesh-prasad-v reviewed (approved):
Nice work on the tests, Kenn! Approved from my side.
